core/src/main/scala/kafka/log/remote/RemoteIndexCache.scala

```diff
@@ -0,0 +1,284 @@
+ * re-fetching the index files like offset, time indexes from the remote storage for every fetch call.
+ *
+ * @param maxSize
+ * @param remoteStorageManager
+ * @param logDir
+ */
+//todo-tier make maxSize configurable
+class RemoteIndexCache(maxSize: Int = 1024, remoteStorageManager: RemoteStorageManager, logDir: String) extends Logging {
+
+  val cacheDir = new File(logDir, DirName)
+  @volatile var closed = false
+
+  val expiredIndexes = new LinkedBlockingQueue[Entry]()
+  val lock = new Object()
+
+  val entries: util.Map[Uuid, Entry] = new java.util.LinkedHashMap[Uuid, Entry](maxSize / 2,
+    0.75f, true) {
+    override def removeEldestEntry(eldest: util.Map.Entry[Uuid, Entry]): Boolean = {
+      if (this.size() > maxSize) {
+        val entry = eldest.getValue
+        // Mark the entries for cleanup, background thread will clean them later.
+        entry.markForCleanup()
+        expiredIndexes.add(entry)
+        true
+      } else {
+        false
+      }
+    }
+  }
+
+  private def init(): Unit = {
+    if (cacheDir.mkdir())
+      info(s"Created $cacheDir successfully")
+
+    // Delete any .deleted files remained from the earlier run of the broker.
+    Files.list(cacheDir.toPath).forEach((path: Path) => {
+      if (path.endsWith(UnifiedLog.DeletedFileSuffix)) {
+        Files.deleteIfExists(path)
+      }
+    })
+
+    Files.list(cacheDir.toPath).forEach((path:Path) => {
+      val pathStr = path.getFileName.toString
+      val name = pathStr.substring(0, pathStr.lastIndexOf("_") + 1)
+
+      // Create entries for each path if all the index files exist.
+      val firstIndex = name.indexOf('_')
+      val offset = name.substring(0, firstIndex).toInt
+      val uuid = Uuid.fromString(name.substring(firstIndex + 1, name.lastIndexOf('_')))
+
+      if(!entries.containsKey(uuid)) {
+        val offsetIndexFile = new File(cacheDir, name + UnifiedLog.IndexFileSuffix)
+        val timestampIndexFile = new File(cacheDir, name + UnifiedLog.TimeIndexFileSuffix)
+        val txnIndexFile = new File(cacheDir, name + UnifiedLog.TxnIndexFileSuffix)
+
+        if (offsetIndexFile.exists() && timestampIndexFile.exists() && txnIndexFile.exists()) {
+
+          val offsetIndex: LazyIndex[OffsetIndex] = {
+            val index = LazyIndex.forOffset(offsetIndexFile, offset, Int.MaxValue, writable = false)
+            index.get.sanityCheck()
+            index
+          }
+
+          val timeIndex: LazyIndex[TimeIndex] = {
+            val index = LazyIndex.forTime(timestampIndexFile, offset, Int.MaxValue, writable = false)
+            index.get.sanityCheck()
+            index
+          }
+
+          val txnIndex: TransactionIndex = {
+            val index = new TransactionIndex(offset, txnIndexFile)
+            index.sanityCheck()
+            index
+          }
+
+          val entry = new Entry(offsetIndex, timeIndex, txnIndex)
+          entries.put(uuid, entry)
```

Comment:
should we lock it?

Reply:
Locking is not really needed here as `init` method is invoked only when RemoteIndexCache instance is getting initialized.